keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,3546 @@
+into a set of MIG devices is not as simple as slicing it up into a set of
+fixed-size, non-oerlapping partitions. Instead, individual MIG devices are defined by the
+set of (possibly overlapping) capacities that they consume from a full GPU. In
+other words, the allocation of one MIG device may nullify the ability to allocate some
+subset of other possible MIG devices (but not all of them). A MIG device is considered
+allocatable so long as the full GPU has enough *unallocated* capacity to satisfy
+its capacity constraints across all dimensions. If any single dimension is
+unavailable, that MIG device cannot be allocated (but others might still be able to).
+
+With his motivating example in mind. We define the following goals and
+non-goals of this KEP.
+
+### Goals
+
+* Introduce the ability for "structured parameters" DRA to allocate both full
+  devices and fixed-size partitions of full devices (across multiple
+  dimensions). Both full devices and their valid set of partitions must be
+  explicitly listed in a ResourceSlice for the scheduler to consider them for
+  allocation.
+
+* Provide abstractions to keep the overall footprint of a ResourceSlice small,
+  such that all full devices (and their partitions) can be represented in a
+  concise way.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+* Allow a user to allocate arbitrarily-sized partitions of a device. In other
+  words, only those partitions explicitly defined by a ResourceSlice will be
+  allocatable by a user. Arbitrary slicing is out of scope for this KEP.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+The basic idea is the following:
+
+1. Introduce a new device type called `PartitionableDevice` which has the same
+   fields as a BasicDevice, plus two more. The first is a field called
+   `Includes` and the second is a field called `ConsumesCapacityFrom`. Both
+   full devices and their paritions are represented as instances of this new
+   `PartitionableDevice` type and are listed right next to one another in the
+   top-level `Devices` list of a `ResourceSlice`.
+
+1. The `Includes` field serves to reference a set of "mixins" that a
+   `PartitionableDevice` can reference to extend the set of attributes and
+   capacities it defines explicitly. The goal being to reduce the overall
+   footprint of each device when a common set of attributes and capacities can
+   be applied. The mixins themselves are introduced as a list of top-level
+   objects directly next to the list of `Devices` inside a `ResourceSlice`.
+   They are not allocatable on their own.
+
+1. The `ConsumesCapacityFrom` field contains a list of *other* devices where the
+   capacity of the current device should be consumed if the scheduler decides
+   to allocate it. This essentially removes that capacity from any referenced
```

Comment:
But other partitions that consume the same capacity can still be allocated, right?

Reply:
What do you mean by the "same" capacity? Once a given capacity is depleted (by the allocation of one or more devices) it is no longer available for other partitions that reference it to consume it.

Reply:
`gpu-0-mig-1g.5gb-0` and `gpu-0-mig-1g.5gb-1`, for example

Reply:
They both pull from the same pool of resources availalable on `gpu-0`, namely:
```
- name: gpu-0
    ...
    capacity:
      copy-engines:
        quantity: "7"
      decoders:
        quantity: "5"
      encoders:
        quantity: "0"
      jpeg-engines:
        quantity: "1"
      memory:
        quantity: 40Gi
      memorySlice0:
        quantity: "1"
      memorySlice1:
        quantity: "1"
      memorySlice2:
        quantity: "1"
      memorySlice3:
        quantity: "1"
      memorySlice4:
        quantity: "1"
      memorySlice5:
        quantity: "1"
      memorySlice6:
        quantity: "1"
      memorySlice7:
        quantity: "1"
      multiprocessors:
        quantity: "98"
      ofa-engines:
        quantity: "1"
```
But both of them can be allocated without consuming pushing any of the capacities available on GPU beyond 0.

Reply:
The text for this has been updated to explain this better.